File: packages/core/src/log.js
import { randomUUID } from 'crypto';
import { debug } from './logger.js';
import { INGEST_PATH } from './constants.js';
export function buildLogEntry(input, result, verdict, userDecision, projectSlug, durationMs) {
    return {
        id: randomUUID(),
        projectSlug,
        actionType: input.actionType,
        verdict: result.verdict,
        userDecision,
        policyTriggered: result.policyName,
        blockReason: result.blockReason,
        durationMs,
        timestamp: new Date().toISOString(),
    };
}
/**
 * Fire-and-forget log sender. Never awaited. Never throws.
 * Sends only behavioral metadata, never actionPayload.
 */
export function sendLogEntry(entry, config) {
    const url = `${config.apiEndpoint}${INGEST_PATH}`;
    fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'x-vouch-key': config.apiKey,
        },
        body: JSON.stringify(entry),
    }).catch((err) => {
        debug('Failed to send log entry', err instanceof Error ? err.message : err);
    });
}

File: packages/sdk-python/vouch/log.py
from __future__ import annotations

import json
import os
import sys
import time
import uuid
from typing import Optional

from .types import ActionLogEntry, PolicyResult, VouchConfig, VouchInput


def build_log_entry(
    input: VouchInput,
    result: PolicyResult,
    verdict: str,
    user_decision: Optional[str],
    project_slug: str,
    duration_ms: int,
) -> ActionLogEntry:
    return ActionLogEntry(
        id=str(uuid.uuid4()),
        project_slug=project_slug,
        action_type=input.action_type,
        verdict=result.verdict,
        user_decision=user_decision,
        policy_triggered=result.policy_name,
        block_reason=result.block_reason,
        duration_ms=duration_ms,
        timestamp=time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    )


def send_log_entry(entry: ActionLogEntry, config: VouchConfig) -> None:
    """Fire-and-forget log sender. Never raises."""
    try:
        import threading

        def _send() -> None:
            try:
                url = f"{config.api_endpoint}/api/ingest"
                payload = json.dumps({
                    "id": entry.id,
                    "projectSlug": entry.project_slug,
                    "actionType": entry.action_type,
                    "verdict": entry.verdict,
                    "userDecision": entry.user_decision,
                    "policyTriggered": entry.policy_triggered,
                    "blockReason": entry.block_reason,
                    "durationMs": entry.duration_ms,
                    "timestamp": entry.timestamp,
                }).encode("utf-8")

                try:
                    import requests
                    requests.post(
                        url,
                        data=payload,
                        headers={
                            "Content-Type": "application/json",
                            "x-vouch-key": config.api_key,
                        },
                        timeout=5,
                    )
                except ImportError:
                    from urllib.request import Request, urlopen
                    req = Request(url, data=payload, method="POST")
                    req.add_header("Content-Type", "application/json")
                    req.add_header("x-vouch-key", config.api_key)
                    urlopen(req, timeout=5)
            except Exception:
                if os.environ.get("VOUCH_DEBUG") == "true":
                    import traceback
                    sys.stderr.write(f"[vouch] Failed to send log: {traceback.format_exc()}\n")

        t = threading.Thread(target=_send, daemon=True)
        t.start()
    except Exception:
        pass
